From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 26 Oct 2023 14:26:41 +0200
Subject: [PATCH] HACK: dmime: Don't send segment end message for secondary
 segments.

CW-Bug-Id: #16680
CW-Bug-Id: #20424
CW-Bug-Id: #22409
Valve-commit: 46cedf9d9fe872e406f6e1ad213dcf91bce0a982
Picked: resolved
Class: valve
Evidence: docs/evidence/2026-09-28-wine-proton-rebase.md
Offered-upstream: no
---
 dlls/dmime/dmime_private.h | 2 +-
 dlls/dmime/performance.c   | 2 +-
 dlls/dmime/segmentstate.c  | 7 +++++--
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/dlls/dmime/dmime_private.h b/dlls/dmime/dmime_private.h
index 728db4f..888d6b2 100644
--- a/dlls/dmime/dmime_private.h
+++ b/dlls/dmime/dmime_private.h
@@ -79,7 +79,7 @@ extern void set_audiopath_primary_dsound_buffer(IDirectMusicAudioPath*,IDirectSo
 
 extern HRESULT segment_get_rt_length(IDirectMusicSegment *iface, REFERENCE_TIME *length);
 
-extern HRESULT segment_state_create(IDirectMusicSegment *segment, MUSIC_TIME start_time,
+extern HRESULT segment_state_create(IDirectMusicSegment *segment, MUSIC_TIME start_time, DWORD segment_flags,
         IDirectMusicPerformance8 *performance, IDirectMusicSegmentState **ret_iface);
 extern HRESULT segment_state_play(IDirectMusicSegmentState *iface, IDirectMusicPerformance8 *performance);
 extern HRESULT segment_state_tick(IDirectMusicSegmentState *iface, IDirectMusicPerformance8 *performance);
diff --git a/dlls/dmime/performance.c b/dlls/dmime/performance.c
index 9ab6e8c..a0c6929 100644
--- a/dlls/dmime/performance.c
+++ b/dlls/dmime/performance.c
@@ -1627,7 +1627,7 @@ static HRESULT WINAPI performance_PlaySegmentEx(IDirectMusicPerformance8 *iface,
     if (control) performance_set_control_segment(This, segment);
 
     if ((!(music_time = start_time) && FAILED(hr = IDirectMusicPerformance8_GetTime(iface, NULL, &music_time)))
-            || FAILED(hr = segment_state_create(segment, music_time, iface, &state)))
+            || FAILED(hr = segment_state_create(segment, music_time, segment_flags, iface, &state)))
     {
         if (primary) performance_set_primary_segment(This, NULL);
         if (control) performance_set_control_segment(This, NULL);
diff --git a/dlls/dmime/segmentstate.c b/dlls/dmime/segmentstate.c
index 32b9600..e363022 100644
--- a/dlls/dmime/segmentstate.c
+++ b/dlls/dmime/segmentstate.c
@@ -62,6 +62,7 @@ struct segment_state
     BOOL auto_download;
     DWORD repeats, actual_repeats;
     DWORD track_flags;
+    DWORD segment_flags;
 
     struct list tracks;
 };
@@ -119,7 +120,7 @@ static ULONG WINAPI segment_state_Release(IDirectMusicSegmentState8 *iface)
 
     if (!ref)
     {
-        segment_state_end_play((IDirectMusicSegmentState *)iface, NULL);
+        if (!(This->segment_flags & DMUS_SEGF_SECONDARY)) segment_state_end_play((IDirectMusicSegmentState *)iface, NULL);
         if (This->segment) IDirectMusicSegment_Release(This->segment);
         if (This->parent_graph) IDirectMusicGraph_Release(This->parent_graph);
         free(This);
@@ -291,7 +292,7 @@ HRESULT create_dmsegmentstate(REFIID riid, void **ret_iface)
     return hr;
 }
 
-HRESULT segment_state_create(IDirectMusicSegment *segment, MUSIC_TIME start_time,
+HRESULT segment_state_create(IDirectMusicSegment *segment, MUSIC_TIME start_time, DWORD segment_flags,
         IDirectMusicPerformance8 *performance, IDirectMusicSegmentState **ret_iface)
 {
     IDirectMusicSegmentState *iface;
@@ -308,6 +309,7 @@ HRESULT segment_state_create(IDirectMusicSegment *segment, MUSIC_TIME start_time
     if (FAILED(hr = create_dmsegmentstate(&IID_IDirectMusicSegmentState, (void **)&iface))) return hr;
     This = impl_from_IDirectMusicSegmentState8((IDirectMusicSegmentState8 *)iface);
 
+    This->segment_flags = segment_flags;
     This->segment = segment;
     IDirectMusicSegment_AddRef(This->segment);
 
@@ -435,6 +437,7 @@ static HRESULT segment_state_play_chunk(struct segment_state *This, IDirectMusic
         {
             MUSIC_TIME end_time = This->actual_start_time + This->played;
 
+            if (This->segment_flags & DMUS_SEGF_SECONDARY) return S_OK;
             if (FAILED(hr = performance_send_segment_end(performance, end_time, iface, FALSE)))
             {
                 ERR("Failed to send segment end, hr %#lx\n", hr);
